server/src/main/java/org/apache/uniffle/server/netty/ShuffleServerNettyHandler.java

```diff
@@ -459,29 +463,27 @@ public void handleGetLocalShuffleData(TransportClient client, GetLocalShuffleDat
                     storageType,
                     offset,
                     length);
-        long readTime = System.currentTimeMillis() - start;
-        ShuffleServerMetrics.counterTotalReadTime.inc(readTime);
         ShuffleServerMetrics.counterTotalReadDataSize.inc(sdr.getDataLength());
         ShuffleServerMetrics.counterTotalReadLocalDataFileSize.inc(sdr.getDataLength());
-        shuffleServer
-            .getNettyMetrics()
-            .recordProcessTime(GetLocalShuffleDataRequest.class.getName(), readTime);
-        LOG.info(
-            "Successfully getShuffleData cost {} ms for shuffle" + " data with {}",
-            readTime,
-            requestInfo);
         response =
             new GetLocalShuffleDataResponse(
                 req.getRequestId(), status, msg, sdr.getManagedBuffer());
+        ReleaseMemoryAndRecordReadTimeListener listener =
+            new ReleaseMemoryAndRecordReadTimeListener(
+                start, length, sdr.getDataLength(), requestInfo, req, client);
+        client.getChannel().writeAndFlush(response).addListener(listener);
+        return;
       } catch (Exception e) {
+        shuffleServer.getShuffleBufferManager().releaseReadMemory(length);
```

Comment:
I have some concern about this. Will it release twice? Because maybe completelistener will release memory when it throws exception.

Reply:
No, `writeAndFlush` is asynchronous, so no exceptions will be catched here.

Reply:
Why do we throw an exception in the future listener?

Reply:
```
throw new RssException("Can not handle request " + request.type());
```
You mean this code snippet? I think we will never reach here, so it's fine we do anything.

Reply:
Yes, this exception will be ignored now. It's not a good way.

Reply:
Or I can just log an error message if you prefer.

Reply:
ERROR log message or assertion is  acceptable. It's up to you.

Reply:
Done.